Build on [Pandas Fundamentals](pandas_fundamentals.ipynb) by transforming labeled tables. No NumPy knowledge is needed for this lesson. Next comes [NumPy Fundamentals](numpy_fundamentals.ipynb), followed by [NumPy Speedup Pandas](numpy_speedup_pandas.ipynb).

## Learning Objectives

- Calculate new columns and explain row versus column summaries.
- Predict label alignment and handle missing values deliberately.
- Choose column expressions, conditional updates, mapping, or custom functions.
- Work with strings, dates, ranks, sorting keys, and correlations.
- Prepare labeled inputs for a later numerical workflow.

[Advanced NLP](#final-project-nlp) is a **final-project resource**, analogous to geospatial plotting in the visualization material. Use it when your project involves text; it remains part of this chapter.


In [1]:
import pandas as pd

## **Arithmetic Operations:** Within and Between DataFrames

Arithmetic operations are essential when working with **numerical data columns** in pandas.  

### Why Arithmetic Operations Matter in Data Science
They form the backbone of many common tasks, including: 

- **Creating new metrics** — e.g., `profit = revenue - cost`  
- **Normalizing data** — z-scores, percentages, or scaling  
- **Comparing datasets** — growth rates, ratios, or relative differences  
- **Financial analysis** — ROI, margins, compound growth  
- **Scientific computations** — applying formulas and transformations  


**🚀 Pandas Advantage:** Arithmetic in pandas is **intuitive and concise** — operations align automatically by rows and columns, saving you from manual looping.

### Arithmetic Operations Within a DataFrame

**Core Concept**: Perform calculations **between columns**, **with constants**, or **across rows**

| Operation | Symbol | Example | Business Use Case |
|-----------|--------|---------|-------------------|
| **Addition** | `+` | `df['A'] + df['B']` | Total sales = Online + Retail |
| **Subtraction** | `-` | `df['Revenue'] - df['Cost']` | Profit calculation |
| **Multiplication** | `*` | `df['Price'] * df['Quantity']` | Total value |
| **Division** | `/` | `df['Profit'] / df['Revenue']` | Margin percentage |
| **Power** | `**` | `df['Principal'] * (1 + rate)**years` | Compound interest |
| **Floor Division** | `//` | `df['Total'] // df['BatchSize']` | Number of full batches |
| **Modulo** | `%` | `df['ID'] % 10` | Grouping by last digit |

#### Column-to-Column Operations

In [2]:
# 🎯 Create a business-like DataFrame for demonstrations
business_data = {
    'Product': ['Laptop', 'Phone', 'Tablet', 'Watch', 'Headphones'],
    'Units_Sold': [150, 300, 200, 500, 250],
    'Price_per_Unit': [1200, 800, 600, 400, 200],
    'Cost_per_Unit': [800, 500, 400, 250, 120],
    'Marketing_Spend': [15000, 25000, 18000, 30000, 12000]
}
business_df = pd.DataFrame(business_data)

print("📊 Original Business Data:")
print(business_df)

# 💰 Basic arithmetic operations between columns
business_df['Total_Revenue'] = business_df['Units_Sold'] * business_df['Price_per_Unit']
business_df['Total_Cost'] = business_df['Units_Sold'] * business_df['Cost_per_Unit']  
business_df['Gross_Profit'] = business_df['Total_Revenue'] - business_df['Total_Cost']
business_df['Net_Profit'] = business_df['Gross_Profit'] - business_df['Marketing_Spend']

# 📈 Calculate ratios and percentages
business_df['Profit_Margin'] = (business_df['Gross_Profit'] / business_df['Total_Revenue']) * 100
business_df['ROI'] = (business_df['Net_Profit'] / business_df['Marketing_Spend']) * 100

print("\n💡 Enhanced Business Data with Calculations:")
print(business_df[['Product', 'Total_Revenue', 'Gross_Profit', 'Net_Profit', 'Profit_Margin', 'ROI']].round(2))



📊 Original Business Data:
      Product  Units_Sold  Price_per_Unit  Cost_per_Unit  Marketing_Spend
0      Laptop         150            1200            800            15000
1       Phone         300             800            500            25000
2      Tablet         200             600            400            18000
3       Watch         500             400            250            30000
4  Headphones         250             200            120            12000

💡 Enhanced Business Data with Calculations:
      Product  Total_Revenue  Gross_Profit  Net_Profit  Profit_Margin     ROI
0      Laptop         180000         60000       45000          33.33  300.00
1       Phone         240000         90000       65000          37.50  260.00
2      Tablet         120000         40000       22000          33.33  122.22
3       Watch         200000         75000       45000          37.50  150.00
4  Headphones          50000         20000        8000          40.00   66.67


#### Operations with Constants and Scalars

In pandas, you can perform arithmetic operations not only between columns, but also with **constants (scalars)**.  
The operation is applied **element-wise** across the entire column or DataFrame — thanks to vectorization.  

In [3]:
# 🎯 Operations with constants (scalars)
print("Operations with Constants:")

# Apply 10% discount to all prices
business_df['Discounted_Price'] = business_df['Price_per_Unit'] * 0.9

# Add fixed shipping cost of $50 to all products
business_df['Price_with_Shipping'] = business_df['Price_per_Unit'] + 50

# Calculate tax (8% of revenue)
business_df['Tax_Amount'] = business_df['Total_Revenue'] * 0.08

# Square the units sold (for some statistical analysis)
business_df['Units_Squared'] = business_df['Units_Sold'] ** 2

print("Sample with scalar operations:")
sample_cols = ['Product', 'Price_per_Unit', 'Discounted_Price', 'Price_with_Shipping', 'Tax_Amount']
print(business_df[sample_cols].head(3).to_string(index=False))

Operations with Constants:
Sample with scalar operations:
Product  Price_per_Unit  Discounted_Price  Price_with_Shipping  Tax_Amount
 Laptop            1200            1080.0                 1250     14400.0
  Phone             800             720.0                  850     19200.0
 Tablet             600             540.0                  650      9600.0


#### Row-wise Operations (Across Columns)

By default, methods like `.mean()`, `.max()`, and `.min()` operate **down each column** (`axis=0`).  
However, if you need to perform calculations **across all columns in each row**, set `axis=1`.

In [4]:
# 📊 Row-wise arithmetic operations (axis=1 for across columns)
print("🔄 Row-wise Operations:")

# Create a sample DataFrame with numerical data
sample_data = {
    'Q1_Sales': [100, 150, 200, 120, 180],
    'Q2_Sales': [110, 140, 220, 130, 190], 
    'Q3_Sales': [120, 160, 210, 140, 200],
    'Q4_Sales': [130, 170, 230, 150, 210]
}
quarterly_sales = pd.DataFrame(sample_data, 
                              index=['Product_A', 'Product_B', 'Product_C', 'Product_D', 'Product_E'])

print("📈 Quarterly Sales Data:")
print(quarterly_sales)

# Calculate row-wise statistics
quarterly_sales['Total_Sales'] = quarterly_sales.sum(axis=1)  # Sum across columns
quarterly_sales['Average_Sales'] = quarterly_sales[['Q1_Sales', 'Q2_Sales', 'Q3_Sales', 'Q4_Sales']].mean(axis=1)
quarterly_sales['Max_Quarter'] = quarterly_sales[['Q1_Sales', 'Q2_Sales', 'Q3_Sales', 'Q4_Sales']].max(axis=1)
quarterly_sales['Sales_Range'] = quarterly_sales[['Q1_Sales', 'Q2_Sales', 'Q3_Sales', 'Q4_Sales']].max(axis=1) - quarterly_sales[['Q1_Sales', 'Q2_Sales', 'Q3_Sales', 'Q4_Sales']].min(axis=1)

print("\n📊 With Row-wise Calculations:")
print(quarterly_sales.round(2))

🔄 Row-wise Operations:
📈 Quarterly Sales Data:
           Q1_Sales  Q2_Sales  Q3_Sales  Q4_Sales
Product_A       100       110       120       130
Product_B       150       140       160       170
Product_C       200       220       210       230
Product_D       120       130       140       150
Product_E       180       190       200       210

📊 With Row-wise Calculations:
           Q1_Sales  Q2_Sales  ...  Max_Quarter  Sales_Range
Product_A       100       110  ...          130           30
Product_B       150       140  ...          170           30
Product_C       200       220  ...          230           30
Product_D       120       130  ...          150           30
Product_E       180       190  ...          210           30

[5 rows x 8 columns]


###  Arithmetic Operations Between DataFrames

**Key Concept**: When operating on **two DataFrames**, pandas **automatically aligns** data by **index** and **column labels**.

#### Alignment Behavior

| Scenario | Result | Example |
|----------|---------|---------|
| **Exact match** | Normal operation | `df1 + df2` |
| **Missing index/column** | Fills with `NaN` | Automatic alignment |
| **Different shapes** | Broadcasts when possible | Scalar operations |
| **Using fill_value** | Replaces NaN with specified value | `df1.add(df2, fill_value=0)` |

**💡 Pro Tip**: Use **explicit methods** (`add()`, `sub()`, `mul()`, `div()`) for better control over missing data!

#### Basic DataFrame-to-DataFrame Operations

In [5]:
# 🏢 Create two related business DataFrames for comparison
print("🔀 DataFrame-to-DataFrame Operations:")

# 📊 Store A performance data
store_a_data = {
    'Electronics': [50000, 55000, 60000],
    'Clothing': [30000, 32000, 35000],
    'Books': [8000, 8500, 9000]
}
store_a = pd.DataFrame(store_a_data, index=['Q1', 'Q2', 'Q3'])

# 📊 Store B performance data  
store_b_data = {
    'Electronics': [45000, 50000, 58000],
    'Clothing': [28000, 30000, 33000],
    'Books': [7500, 8000, 8800]
}
store_b = pd.DataFrame(store_b_data, index=['Q1', 'Q2', 'Q3'])

print("🏪 Store A Sales:")
print(store_a)
print("\n🏪 Store B Sales:")
print(store_b)

# 📈 Basic arithmetic operations between DataFrames
print("\n💡 Arithmetic Operations Between Stores:")

# Total sales across both stores
total_sales = store_a + store_b
print("\n📊 Combined Sales (A + B):")
print(total_sales)

# Difference in performance
sales_diff = store_a - store_b  
print("\n📊 Sales Difference (A - B):")
print(sales_diff)

# Growth ratio (Store A relative to Store B)
growth_ratio = store_a / store_b
print("\n📊 Performance Ratio (A / B):")
print(growth_ratio.round(3))

🔀 DataFrame-to-DataFrame Operations:
🏪 Store A Sales:
    Electronics  Clothing  Books
Q1        50000     30000   8000
Q2        55000     32000   8500
Q3        60000     35000   9000

🏪 Store B Sales:
    Electronics  Clothing  Books
Q1        45000     28000   7500
Q2        50000     30000   8000
Q3        58000     33000   8800

💡 Arithmetic Operations Between Stores:

📊 Combined Sales (A + B):
    Electronics  Clothing  Books
Q1        95000     58000  15500
Q2       105000     62000  16500
Q3       118000     68000  17800

📊 Sales Difference (A - B):
    Electronics  Clothing  Books
Q1         5000      2000    500
Q2         5000      2000    500
Q3         2000      2000    200

📊 Performance Ratio (A / B):
    Electronics  Clothing  Books
Q1        1.111     1.071  1.067
Q2        1.100     1.067  1.062
Q3        1.034     1.061  1.023


#### Handling Misaligned Data

In [6]:
# 🚨 Demonstrate alignment issues and solutions
print("⚠️ Handling Misaligned DataFrames:")

# Create DataFrames with different indices and columns
df1_partial = pd.DataFrame({
    'A': [10, 20, 30],
    'B': [40, 50, 60]
}, index=[0, 1, 2])

df2_partial = pd.DataFrame({
    'B': [5, 10, 15, 20],  # Extra row
    'C': [100, 200, 300, 400]  # Different column
}, index=[1, 2, 3, 4])  # Different indices

print("📊 DataFrame 1:")
print(df1_partial)
print("\n📊 DataFrame 2:")  
print(df2_partial)

# 🔄 Default behavior - creates NaN for missing alignments
print("\n❌ Default Addition (creates NaN):")
result_nan = df1_partial + df2_partial
print(result_nan)

# 🔧 Using explicit methods with fill_value to handle NaN
print("\n✅ Addition with fill_value=0:")
result_filled = df1_partial.add(df2_partial, fill_value=0)
print(result_filled)

# 🎯 Other explicit arithmetic methods
print("\n🔧 Explicit Methods Available:")
print("• add() - Addition with fill_value option")
print("• sub() - Subtraction with fill_value option")  
print("• mul() - Multiplication with fill_value option")
print("• div() - Division with fill_value option")

# Example with subtraction
result_sub = df1_partial.sub(df2_partial, fill_value=1)
print("\n➖ Subtraction with fill_value=1:")
print(result_sub)

⚠️ Handling Misaligned DataFrames:
📊 DataFrame 1:
    A   B
0  10  40
1  20  50
2  30  60

📊 DataFrame 2:
    B    C
1   5  100
2  10  200
3  15  300
4  20  400

❌ Default Addition (creates NaN):
    A     B   C
0 NaN   NaN NaN
1 NaN  55.0 NaN
2 NaN  70.0 NaN
3 NaN   NaN NaN
4 NaN   NaN NaN

✅ Addition with fill_value=0:
      A     B      C
0  10.0  40.0    NaN
1  20.0  55.0  100.0
2  30.0  70.0  200.0
3   NaN  15.0  300.0
4   NaN  20.0  400.0

🔧 Explicit Methods Available:
• add() - Addition with fill_value option
• sub() - Subtraction with fill_value option
• mul() - Multiplication with fill_value option
• div() - Division with fill_value option

➖ Subtraction with fill_value=1:
      A     B      C
0   9.0  39.0    NaN
1  19.0  45.0  -99.0
2  29.0  50.0 -199.0
3   NaN -14.0 -299.0
4   NaN -19.0 -399.0


###  Advanced Arithmetic Techniques

#### Broadcasting and Shape Compatibility

In [7]:
# 📐 Broadcasting: Operations between different shapes
print("📐 Broadcasting Examples:")

# DataFrame with different dimensions
sales_matrix = pd.DataFrame({
    'Jan': [1000, 1500, 2000],
    'Feb': [1100, 1600, 2100], 
    'Mar': [1200, 1700, 2200]
}, index=['Store_A', 'Store_B', 'Store_C'])

# Series for operations
monthly_bonus = pd.Series([100, 150, 200], index=['Jan', 'Feb', 'Mar'])  # Column-wise
store_multiplier = pd.Series([1.1, 1.2, 1.05], index=['Store_A', 'Store_B', 'Store_C'])  # Row-wise

print("📊 Original Sales Matrix:")
print(sales_matrix)
print("\n💰 Monthly Bonus (Series):")
print(monthly_bonus)
print("\n📈 Store Multiplier (Series):")
print(store_multiplier)

# 🔄 Broadcasting operations
print("\n✨ Broadcasting Results:")

# Add monthly bonus to each column
sales_with_bonus = sales_matrix + monthly_bonus
print("📊 Sales + Monthly Bonus (broadcasting across columns):")
print(sales_with_bonus)

# Multiply each row by store multiplier
sales_adjusted = sales_matrix.mul(store_multiplier, axis=0)  # axis=0 for row-wise
print("\n📊 Sales × Store Multiplier (broadcasting across rows):")
print(sales_adjusted.round(0))

📐 Broadcasting Examples:
📊 Original Sales Matrix:
          Jan   Feb   Mar
Store_A  1000  1100  1200
Store_B  1500  1600  1700
Store_C  2000  2100  2200

💰 Monthly Bonus (Series):
Jan    100
Feb    150
Mar    200
dtype: int64

📈 Store Multiplier (Series):
Store_A    1.10
Store_B    1.20
Store_C    1.05
dtype: float64

✨ Broadcasting Results:
📊 Sales + Monthly Bonus (broadcasting across columns):
          Jan   Feb   Mar
Store_A  1100  1250  1400
Store_B  1600  1750  1900
Store_C  2100  2250  2400

📊 Sales × Store Multiplier (broadcasting across rows):
            Jan     Feb     Mar
Store_A  1100.0  1210.0  1320.0
Store_B  1800.0  1920.0  2040.0
Store_C  2100.0  2205.0  2310.0


## Conditional Updates and Missing Values

Use masks to state which records qualify. `.where(condition, other)` keeps values where the condition is true; `.mask(condition, other)` replaces those values. Missing or invalid denominators should remain missing unless the question supplies a defensible replacement.


In [8]:
orders = pd.DataFrame({'price': [20., 50., 30., 10.], 'quantity': [2, 1, 3, 0],
                       'discount': [0.1, 0.2, 0., 0.]}, index=[104, 101, 109, 103])
orders['total'] = orders['price'] * orders['quantity'] * (1 - orders['discount'])
orders['size'] = 'Regular'
orders.loc[orders['total'] >= 50, 'size'] = 'Large'
orders['unit_total'] = orders['total'].div(orders['quantity'].where(orders['quantity'] != 0))
orders['capped_total'] = orders['total'].mask(orders['total'] > 60, 60)
print(orders)
assert pd.isna(orders.loc[103, 'unit_total'])


     price  quantity  discount  total     size  unit_total  capped_total
104   20.0         2       0.1   36.0  Regular        18.0          36.0
101   50.0         1       0.2   40.0  Regular        40.0          40.0
109   30.0         3       0.0   90.0    Large        30.0          60.0
103   10.0         0       0.0    0.0  Regular         NaN           0.0


## Mapping, Replacement, and Functions

A dictionary or Series passed to `map()` describes a lookup. Unmapped values become missing. `replace()` changes specified values and keeps other values. A function describes a computation; define it before passing it to `map()` or `apply()`.


In [9]:
status = pd.Series(['Agree', 'Disagree', 'Unsure', None], dtype='string')
print(pd.DataFrame({'original': status,
                    'mapped': status.map({'Agree': 1, 'Disagree': 0}),
                    'replaced': status.replace({'Agree': 'Yes', 'Disagree': 'No'})}))

def describe_total(value):
    if value >= 50:
        return 'Large'
    return 'Regular'
print(orders['total'].map(describe_total))


   original  mapped replaced
0     Agree     1.0      Yes
1  Disagree     0.0       No
2    Unsure     NaN   Unsure
3      <NA>     NaN     <NA>
104    Regular
101    Regular
109      Large
103    Regular
Name: total, dtype: str


### Lambda Functions and `apply()`

A lambda is a function written as one expression. It is useful for short transformations, not a speed optimization. For `DataFrame.apply`, `axis=1` passes each row to the function and `axis=0` passes each column. Prefer a direct column expression for simple arithmetic.


In [10]:
def calculate_total(row):
    return row['price'] * row['quantity'] * (1 - row['discount'])
reference = orders.apply(calculate_total, axis=1)
inline = orders.apply(lambda row: row['price'] * row['quantity'] * (1 - row['discount']), axis=1)
pd.testing.assert_series_equal(reference, inline)
pd.testing.assert_series_equal(reference, orders['total'], check_names=False)
print(reference)
print(orders[['price', 'quantity']].apply(lambda col: col.max() - col.min(), axis=0))


104    36.0
101    40.0
109    90.0
103     0.0
dtype: float64
price       40.0
quantity     3.0
dtype: float64


## Further Work with Selection and Transformation {#fundamentals-extensions}

These worked extensions continue [Pandas Fundamentals](pandas_fundamentals.ipynb#optional-extensions). Read the function and lambda sections above before custom sorting. Each subsection defines its own small dataset and can be run independently with pandas installed; it needs no downloaded data. Predict the output, run it, and explain the result before changing the example.

### Ranking Conventions and Percentiles {#ranking-and-percentiles}

When values tie, a rank requires a convention. Compare all five methods on the same four scores. Missing scores remain unranked with the default missing-value option.


In [11]:
import pandas as pd
extension_scores = pd.Series([90, 90, 70, 50, None], index=['Ada', 'Ben', 'Cy', 'Dee', 'Eli'])
rank_comparison = pd.DataFrame({'score': extension_scores})
for method in ['average', 'min', 'max', 'first', 'dense']:
    rank_comparison[method] = extension_scores.rank(ascending=False, method=method)
rank_comparison


Out[11]: 
     score  average  min  max  first  dense
Ada   90.0      1.5  1.0  2.0    1.0    1.0
Ben   90.0      1.5  1.0  2.0    2.0    1.0
Cy    70.0      3.0  3.0  3.0    3.0    2.0
Dee   50.0      4.0  4.0  4.0    4.0    3.0
Eli    NaN      NaN  NaN  NaN    NaN    NaN


`average` gives the tied leaders 1.5; `min` gives them 1; `max` gives them 2. `first` breaks the tie using current row order, so it can change after sorting. `dense` assigns consecutive ranks to distinct scores, leaving no gap after a tie.

With the default average method and `pct=True`, pandas divides each rank by the number of observed scores. Choose ascending ranks if larger scores should receive larger fractions. This is a rank convention, not a statement that a certain percentage scored strictly below the student.


In [12]:
percentile_example = pd.DataFrame({'score': extension_scores})
percentile_example['ascending_rank'] = extension_scores.rank(method='average')
percentile_example['rank_fraction'] = extension_scores.rank(method='average', pct=True)
percentile_example


Out[12]: 
     score  ascending_rank  rank_fraction
Ada   90.0             3.5          0.875
Ben   90.0             3.5          0.875
Cy    70.0             2.0          0.500
Dee   50.0             1.0          0.250
Eli    NaN             NaN            NaN


There are four observed scores. Ada and Ben share ascending rank 3.5, so each receives 3.5/4 = 0.875. Different tie methods produce different rank fractions; with `dense`, the denominator is the number of distinct observed values.

**Practice:** Reverse the row order and compare `first` with `average`. Then compare selecting the two largest scores with selecting every row whose competition rank is at most two. Describe the tie policy you would use for an award.

Reference: [pandas rank](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rank.html).

### Custom Sorting Keys {#custom-sorting-keys}

A sorting key computes temporary comparison values while keeping the original values in the result. The function receives a Series and must return a same-length result. A lambda is simply a short way to define that function.


In [13]:
import pandas as pd
extension_titles = pd.DataFrame({'title': ['cedar', 'Harbor', 'ember', 'Meadow']}, index=[104, 101, 109, 103])
case_order = extension_titles.sort_values('title', key=lambda col: col.str.casefold())
length_order = extension_titles.sort_values('title', key=lambda col: col.str.len(), kind='stable')
print('Case-insensitive order:')
print(case_order)
print('\nShortest titles first:')
print(length_order)
print('\nOriginal labels in label order:')
print(extension_titles.sort_index())


Case-insensitive order:
      title
104   cedar
109   ember
101  Harbor
103  Meadow

Shortest titles first:
      title
104   cedar
109   ember
101  Harbor
103  Meadow

Original labels in label order:
      title
101  Harbor
103  Meadow
104   cedar
109   ember


The stored capitalization is unchanged. Equal lengths retain their input order in the stable single-column sort. For different transformations on different columns, create named helper columns and sort by those columns; a key is applied independently to every selected sorting column.

**Practice:** Add a second five-letter title. Predict its location under a stable length sort, then create a helper for a case-insensitive alphabetical tie-breaker.

### Extended Strings, Dates, and Durations {#extended-strings-and-dates}

Inspect representative text and missing values before cleaning. Preserve original values so that a transformation can be checked. Literal replacement is appropriate for a known character; regular expressions are useful when you deliberately need a pattern.


In [14]:
import pandas as pd
extension_text = pd.Series(['  North Hall ', 'SOUTH-HALL', None, 'East Hall'], dtype='string')
text_report = pd.DataFrame({'original': extension_text})
text_report['clean'] = extension_text.str.strip().str.casefold().str.replace('-', ' ', regex=False)
text_report['characters'] = text_report['clean'].str.len()
text_report['hall'] = text_report['clean'].str.endswith('hall', na=False)
text_report


Out[14]: 
        original       clean  characters   hall
0    North Hall   north hall          10   True
1     SOUTH-HALL  south hall          10   True
2           <NA>        <NA>        <NA>  False
3      East Hall   east hall           9   True


Missing text remains missing during cleaning; the explicit `na=False` classifies it as not matching the suffix. `str.split()` can separate fields, and `str.extract()` can capture a regular-expression group; inspect unmatched rows before using extracted values.

Dates need both a parsing rule and a reference frame. Calendar components describe a date, while subtracting two dates measures elapsed time.


In [15]:
import pandas as pd
extension_raw_dates = pd.Series(['2024-01-01', '2024-03-31', '2024-12-30', 'invalid', None])
extension_dates = pd.to_datetime(extension_raw_dates, format='%Y-%m-%d', errors='coerce')
date_report = pd.DataFrame({'original': extension_raw_dates, 'date': extension_dates})
date_report['year'] = extension_dates.dt.year
date_report['month'] = extension_dates.dt.month
date_report['quarter'] = extension_dates.dt.quarter
date_report['weekday'] = extension_dates.dt.day_name()
date_report['iso_year'] = extension_dates.dt.isocalendar().year
date_report['iso_week'] = extension_dates.dt.isocalendar().week
date_report['days_until_2025'] = (pd.Timestamp('2025-01-01') - extension_dates).dt.days
print(date_report)
print('\nObserved text that failed parsing:')
print(extension_raw_dates.loc[extension_raw_dates.notna() & extension_dates.isna()])


     original       date    year  ...  iso_year  iso_week days_until_2025
0  2024-01-01 2024-01-01  2024.0  ...      2024         1           366.0
1  2024-03-31 2024-03-31  2024.0  ...      2024        13           276.0
2  2024-12-30 2024-12-30  2024.0  ...      2025         1             2.0
3     invalid        NaT     NaN  ...      <NA>      <NA>             NaN
4         NaN        NaT     NaN  ...      <NA>      <NA>             NaN

[5 rows x 9 columns]

Observed text that failed parsing:
3    invalid
dtype: str


December 30, 2024 falls in ISO week 1 of ISO year 2025; calendar year and ISO year need not match. A negative duration would mean a date falls after the reference date. For timestamps with time zones, establish compatible zones before subtraction; elapsed hours and local calendar days can differ around clock changes.

**Practice:** Add `2025-01-03`, predict its duration, and compare `.dt.day` (day of month) with `.dt.dayofweek` (Monday is 0). Explain why `invalid` and the originally missing entry both yield missing date features but represent different data-quality situations.



### Copies and Reliable Updates {#performance-and-copies}

Use `.loc` for updates to a particular DataFrame. An alias names the same object; an explicit copy makes independence clear. `inplace=True` is not a guarantee of improved speed or reduced memory. Compare performance only after checking equivalent results.


In [16]:
import pandas as pd
copy_original = pd.DataFrame({'count': [2, 4, 6]})
copy_independent = copy_original.copy()
copy_independent.loc[0, 'count'] = 99
print('Original after changing the copy:', copy_original['count'].tolist())
copy_alias = copy_original
copy_alias.loc[0, 'count'] = 7
print('Original after changing an alias:', copy_original['count'].tolist())
print('Reported column/index bytes:', copy_original.memory_usage(deep=True).sum())


Original after changing the copy: [2, 4, 6]
Original after changing an alias: [7, 4, 6]
Reported column/index bytes: 156


## Correlation

Pearson correlation describes linear association, not causation. Inspect a scatter plot as well as a coefficient. Pandas aligns Series by labels and uses pairs with observed values. The following example uses a small fixed table so the calculation can be checked without random generation. Spearman and Kendall correlations are extensions for rank association.


In [17]:
observations = pd.DataFrame({'advertising': [2., 4., 3., 6., 5.],
                             'sales': [20., 28., 25., 35., 31.],
                             'returns': [1., 3., 2., 2., 4.]}, index=['A','B','C','D','E'])
print(observations.corr())
print(observations['advertising'].corr(observations['sales']))


             advertising     sales  returns
advertising     1.000000  0.995402  0.55470
sales           0.995402  1.000000  0.55215
returns         0.554700  0.552150  1.00000
0.9954022744967957


## Practice: Prepare Survey Data

Work through the existing survey transformations below, then explain how unknown answers differ from a recorded disagreement. The mappings should preserve that distinction.


### Practice exercise 1

#### 
Read the file *STAT303-1 survey for data analysis.csv*.

In [18]:
survey_data = pd.read_csv('./Datasets/STAT303-1 survey for data analysis.csv')
print("Survey data loaded successfully!")
print(f"Shape: {survey_data.shape}")
survey_data.head()

Survey data loaded successfully!
Shape: (192, 51)
Out[18]: 
                      Timestamp  ... much_effort_is_lack_of_talent
0   2022/09/13 1:43:34 pm GMT-5  ...                      Disagree
1   2022/09/13 5:28:17 pm GMT-5  ...                      Disagree
2   2022/09/13 7:56:38 pm GMT-5  ...                      Disagree
3  2022/09/13 10:34:37 pm GMT-5  ...                      Disagree
4   2022/09/14 4:46:19 pm GMT-5  ...                      Disagree

[5 rows x 51 columns]


#### 
Rename all the columns of the data, except the first two columns, with the shorter names in the list `new_col_names` given below. The order of column names in the list is the same as the order in which the columns are to be renamed starting with the third column from the left.

In [19]:
new_col_names = ['parties_per_month', 'do_you_smoke', 'weed', 'are_you_an_introvert_or_extrovert', 'love_first_sight', 'learning_style', 'left_right_brained', 'personality_type', 'social_media', 'num_insta_followers', 'streaming_platforms', 'expected_marriage_age', 'expected_starting_salary', 'fav_sport', 'minutes_ex_per_week', 'sleep_hours_per_day', 'how_happy', 'farthest_distance_travelled', 'fav_number', 'fav_letter', 'internet_hours_per_day', 'only_child', 'birthdate_odd_even', 'birth_month', 'fav_season', 'living_location_on_campus', 'major', 'num_majors_minors', 'high_school_GPA', 'NU_GPA', 'age', 'height', 'height_father', 'height_mother', 'school_year', 'procrastinator', 'num_clubs', 'student_athlete', 'AP_stats', 'used_python_before', 'dominant_hand', 'childhood_in_US', 'gender', 'region_of_residence', 'political_affliation', 'cant_change_math_ability', 'can_change_math_ability', 'math_is_genetic', 'much_effort_is_lack_of_talent']

In [20]:
survey_data.columns = list(survey_data.columns[0:2])+new_col_names

#### 
Rename the following columns again:

1. Rename `do_you_smoke` to `smoke`.

2. Rename `are_you_an_introvert_or_extrovert` to `introvert_extrovert`.

**Hint:** Use the function [rename()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html)

In [21]:

# Rename specific columns using rename function

survey_data = survey_data.rename(columns={

    'do_you_smoke': 'smoke',

    'are_you_an_introvert_or_extrovert': 'introvert_extrovert'})

print(f"Updated columns: {list(survey_data.columns)}")
print("Specific columns renamed successfully!")

Updated columns: ['Timestamp', 'fav_alcohol', 'parties_per_month', 'smoke', 'weed', 'introvert_extrovert', 'love_first_sight', 'learning_style', 'left_right_brained', 'personality_type', 'social_media', 'num_insta_followers', 'streaming_platforms', 'expected_marriage_age', 'expected_starting_salary', 'fav_sport', 'minutes_ex_per_week', 'sleep_hours_per_day', 'how_happy', 'farthest_distance_travelled', 'fav_number', 'fav_letter', 'internet_hours_per_day', 'only_child', 'birthdate_odd_even', 'birth_month', 'fav_season', 'living_location_on_campus', 'major', 'num_majors_minors', 'high_school_GPA', 'NU_GPA', 'age', 'height', 'height_father', 'height_mother', 'school_year', 'procrastinator', 'num_clubs', 'student_athlete', 'AP_stats', 'used_python_before', 'dominant_hand', 'childhood_in_US', 'gender', 'region_of_residence', 'political_affliation', 'cant_change_math_ability', 'can_change_math_ability', 'math_is_genetic', 'much_effort_is_lack_of_talent']
Specific columns renamed successfully!

#### 
Examine the column `num_insta_followers`. Some numbers in the column contain a comma(`,`) or a tilde(`~`). Remove both these characters from the numbers in the column. 

**Hint:** You may use the function [str.replace()](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.replace.html) of the Pandas Series class.

In [22]:

survey_data_insta = survey_data.copy()
survey_data_insta['num_insta_followers']=survey_data_insta['num_insta_followers'].str.replace(',','')
survey_data_insta['num_insta_followers']=survey_data_insta['num_insta_followers'].str.replace('~','')

In [23]:
survey_data_insta.head()

Out[23]: 
                      Timestamp  ... much_effort_is_lack_of_talent
0   2022/09/13 1:43:34 pm GMT-5  ...                      Disagree
1   2022/09/13 5:28:17 pm GMT-5  ...                      Disagree
2   2022/09/13 7:56:38 pm GMT-5  ...                      Disagree
3  2022/09/13 10:34:37 pm GMT-5  ...                      Disagree
4   2022/09/14 4:46:19 pm GMT-5  ...                      Disagree

[5 rows x 51 columns]


#### 
Use the cleaned dataset where the column `num_insta_followers` has been updated.

The last four variables in the dataset are: 

1. `cant_change_math_ability`

2. `can_change_math_ability`

3. `math_is_genetic`

4. `much_effort_is_lack_of_talent`

Each of the above variables has values - `Agree` / `Disagree`. Replace `Agree` with `1` and `Disagree` with `0`.

**Hint :** You can do it with any one of the following methods:

1. Use the map() function

2. Use the apply() function with the lambda function




In [24]:
# implement a function to convert the column to numeric, forcing errors to NaN

# Define the columns to convert
columns_to_convert = ['cant_change_math_ability', 'can_change_math_ability', 
                      'math_is_genetic', 'much_effort_is_lack_of_talent']

# Method 1: Using map() function (one mapping per column)
for col in columns_to_convert:
    survey_data_insta[col] = survey_data_insta[col].map({'Agree': 1, 'Disagree': 0})

print("Method 1: Using map() function")
print(survey_data_insta[columns_to_convert].head())
print("\nData types:")
print(survey_data_insta[columns_to_convert].dtypes)

# Method 2: Using apply() function with lambda (one mapping per column)
# First, let's reload the data to show this method
survey_data_insta2 = survey_data.copy()
survey_data_insta2['num_insta_followers'] = survey_data_insta2['num_insta_followers'].str.replace(',','')
survey_data_insta2['num_insta_followers'] = survey_data_insta2['num_insta_followers'].str.replace('~','')

for col in columns_to_convert:
    survey_data_insta2[col] = survey_data_insta2[col].apply(lambda x: {'Agree': 1, 'Disagree': 0}.get(x))

print("\n\nMethod 2: Using apply() with lambda function")
print(survey_data_insta2[columns_to_convert].head())



Method 1: Using map() function
   cant_change_math_ability  ...  much_effort_is_lack_of_talent
0                         0  ...                              0
1                         0  ...                              0
2                         0  ...                              0
3                         0  ...                              0
4                         1  ...                              0

[5 rows x 4 columns]

Data types:
cant_change_math_ability         int64
can_change_math_ability          int64
math_is_genetic                  int64
much_effort_is_lack_of_talent    int64
dtype: object


Method 2: Using apply() with lambda function
   cant_change_math_ability  ...  much_effort_is_lack_of_talent
0                         0  ...                              0
1                         0  ...                              0
2                         0  ...                              0
3                         0  ...                              0
4            

## Practice: Shopping with Labeled Data

Read the quantities and prices, explicitly put products in the same order, and calculate each person's total at each store. This is a useful pandas solution already. In [NumPy Speedup Pandas](numpy_speedup_pandas.ipynb#shopping-workflow), revisit the same task as a matrix operation.


In [25]:
food = pd.read_csv('Datasets/food_quantity.csv').set_index('Person')
prices = pd.read_csv('Datasets/price.csv').set_index('Item')
items = ['roll', 'bun', 'cake', 'bread']
quantities = food.loc[:, items]
prices = prices.loc[items, ['Target', 'Kroger']]
costs = pd.DataFrame({store: quantities.mul(prices[store], axis='columns').sum(axis=1)
                      for store in prices.columns})
print(costs)
print(costs.idxmin(axis=1))


         Target  Kroger
Person                 
Ben        50.0    49.0
Barbara    58.5    61.0
Beth       43.5    43.5
Person
Ben        Kroger
Barbara    Target
Beth       Target
dtype: str


**Your turn:** Reverse the rows of the price table before using `.loc[items]`. Explain why the result stays unchanged. Find all cheapest stores if there is a tie. Explain what you would do if a price were missing; a sum that silently skips it would understate the bill.

## Before You Move On

Explain the difference between a row label and its position, why a lookup can produce missing values, and when a column expression replaces a row function. Continue to [NumPy Fundamentals](numpy_fundamentals.ipynb).

## Final Project Resource: Advanced NLP {#final-project-nlp}

Use this resource for projects involving reviews, documents, or other text, just as you would use geospatial plotting for a location-based project. It covers regular expressions, tokenization, stop words, stemming, lemmatization, sentiment, and named entities. These examples need their own package/data setup and some use live web data. Run this section separately after completing the core lesson.


### Advanced NLP Preprocessing

For text data stored in **object (string) columns**, pandas works seamlessly with Python’s built-in string functions.  
Beyond that, you can also use the **`re` (regular expressions)** module and the **NLTK (Natural Language Toolkit)** library for more advanced text preprocessing.  

These tools support text-based final projects and are useful for tasks such as text cleaning, tokenization, and feature extraction — especially if you choose an NLP-focused topic for your **final project**.  

Below, we’ll demonstrate how to use the `re` module and NLTK with pandas columns.

### the `re` module in Python is used for regular expression

The `re` module in Python is used for **regular expressions**, which are powerful tools for text analysis and manipulation. Regular expressions allow you to search, match, and manipulate strings based on specific patterns.

Common Use Cases of `re` in String Text Analysis

* Finding patterns in text: `re.search(r'\d{4}-\d{2}-\d{2}', text)`   # searches for a date in the format YYYY-MM-DD
* Extracting Specific parts of a string: `re.findall(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b', text)`  #extracts email addresses from the text.
* Replacing Parts of a String: `re.sub(r'\$\d+', '[price]', text)`    #replacing any price formatted as $<number> with [price].

##### Commonly Used Patterns in `re`

- **`\d`**: Matches any digit (0-9).
- **`\w`**: Matches any alphanumeric character (letters and numbers).
- **`\s`**: Matches any whitespace character (spaces, tabs, newlines).
- **`[a-z]`**: Matches any lowercase letter from `a` to `z`.
- **`[A-Z]`**: Matches any uppercase letter from `A` to `Z`.
- **`*`**: Matches 0 or more occurrences of the preceding character.
- **`+`**: Matches 1 or more occurrences of the preceding character.
- **`?`**: Matches 0 or 1 occurrence of the preceding character.
- **`^`**: Matches the beginning of a string.
- **`$`**: Matches the end of a string.
- **`|`**: Acts as an OR operator.

In [29]:
import requests

def read_html_tables_from_url(url, match=None, user_agent=None, timeout=10, **kwargs):
    """
    Read HTML tables from a website URL with error handling and customizable options.
    
    Parameters:
    -----------
    url : str
        The website URL to scrape tables from
    match : str, optional
        String or regex pattern to match tables (passed to pd.read_html)
    user_agent : str, optional
        Custom User-Agent header. If None, uses a default Chrome User-Agent
    timeout : int, default 10
        Request timeout in seconds
    **kwargs : dict
        Additional arguments passed to pd.read_html()
    
    Returns:
    --------
    list of DataFrames or None
        List of pandas DataFrames (one per table found), or None if error occurs
    
    Examples:
    ---------
    # Basic usage
    tables = read_html_tables_from_url("https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal)_per_capita")
    
    # Filter tables containing specific text
    tables = read_html_tables_from_url(url, match="Country")
    
    # With custom user agent
    tables = read_html_tables_from_url(url, user_agent="MyBot/1.0")
    """
    import pandas as pd
    import requests
    
    # Default User-Agent if none provided
    if user_agent is None:
        user_agent = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    
    headers = {'User-Agent': user_agent}
    
    try:
        print(f"🌐 Fetching data from: {url}")
        
        # Make the HTTP request
        response = requests.get(url, headers=headers, timeout=timeout)
        response.raise_for_status()  # Raise an exception for bad status codes (4xx or 5xx)
        
        print(f"✅ Successfully retrieved webpage (Status: {response.status_code})")
        
        # Parse HTML tables
        read_html_kwargs = {'match': match} if match else {}
        read_html_kwargs.update(kwargs)  # Add any additional arguments
        
        tables = pd.read_html(response.content, **read_html_kwargs)
        
        print(f"📊 Found {len(tables)} table(s) on the page")
        
        # Display table information
        if tables:
            print(f"\n📋 Table Overview:")
            for i, table in enumerate(tables):
                print(f"  Table {i}: {table.shape} (rows × columns)")
                if i >= 4:  # Limit output for readability
                    print(f"  ... and {len(tables) - 5} more tables")
                    break
        
        return tables
        
    except requests.exceptions.HTTPError as err:
        print(f"❌ HTTP Error: {err}")
        return None
    except requests.exceptions.ConnectionError as err:
        print(f"❌ Connection Error: {err}")
        return None
    except requests.exceptions.Timeout as err:
        print(f"❌ Timeout Error: Request took longer than {timeout} seconds")
        return None
    except requests.exceptions.RequestException as err:
        print(f"❌ Request Error: {err}")
        return None
    except ValueError as err:
        print(f"❌ No tables found on the page or parsing error: {err}")
        return None
    except Exception as e:
        print(f"❌ An unexpected error occurred: {e}")
        return None

In [33]:
# Example 1: Basic usage - Get all tables from GDP Wikipedia page
url = 'https://en.wikipedia.org/wiki/List_of_Chicago_Bulls_seasons'
tables = read_html_tables_from_url(url)
ChicagoBulls = tables[2] if tables else None
ChicagoBulls.head()


🌐 Fetching data from: https://en.wikipedia.org/wiki/List_of_Chicago_Bulls_seasons
✅ Successfully retrieved webpage (Status: 200)
📊 Found 18 table(s) on the page

📋 Table Overview:
  Table 0: (13, 2) (rows × columns)
  Table 1: (1, 5) (rows × columns)
  Table 2: (59, 13) (rows × columns)
  Table 3: (3, 4) (rows × columns)
  Table 4: (9, 2) (rows × columns)
  ... and 13 more tables
✅ Successfully retrieved webpage (Status: 200)
📊 Found 18 table(s) on the page

📋 Table Overview:
  Table 0: (13, 2) (rows × columns)
  Table 1: (1, 5) (rows × columns)
  Table 2: (59, 13) (rows × columns)
  Table 3: (3, 4) (rows × columns)
  Table 4: (9, 2) (rows × columns)
  ... and 13 more tables


,Season,Team,Conference,Finish,Division,Finish.1,Wins,Losses,Win%,GB,Playoffs,Awards,Head coach
0,1966–67,1966–67,—,—,Western,4th,33,48,0.407,11,Lost Division semifinals (Hawks) 3–0[19],Johnny Kerr (COY)[6],Johnny Kerr
1,1967–68,1967–68,—,—,Western,4th,29,53,0.354,27,Lost Division semifinals (Lakers) 4–1[20],—,Johnny Kerr
2,1968–69,1968–69,—,—,Western,5th,33,49,0.402,22,NaN,—,Dick Motta
3,1969–70,1969–70,—,—,Western,3rd[c],39,43,0.476,9,Lost Division semifinals (Hawks) 4–1[22],—,Dick Motta
4,1970–71,1970–71,Western,3rd,Midwest[d],2nd,51,31,0.622,2,Lost conference semifinals (Lakers) 4–3[23],Dick Motta (COY)[6],Dick Motta


In [34]:
# remove all charaters between box brackets inluding the brackets themselves in the columns Division, Finish, Playoffs, Awards
import re
def remove_brackets(x):
    return re.sub(r'\[.*?\]', '', x)

# Apply the function to each column separately using map
ChicagoBulls['Division'] = ChicagoBulls['Division'].map(remove_brackets)
ChicagoBulls['Finish'] = ChicagoBulls['Finish'].map(remove_brackets)
ChicagoBulls['Finish.1'] = ChicagoBulls['Finish.1'].map(remove_brackets)

ChicagoBulls.head()

,Season,Team,Conference,Finish,Division,Finish.1,Wins,Losses,Win%,GB,Playoffs,Awards,Head coach
0,1966–67,1966–67,—,—,Western,4th,33,48,0.407,11,Lost Division semifinals (Hawks) 3–0[19],Johnny Kerr (COY)[6],Johnny Kerr
1,1967–68,1967–68,—,—,Western,4th,29,53,0.354,27,Lost Division semifinals (Lakers) 4–1[20],—,Johnny Kerr
2,1968–69,1968–69,—,—,Western,5th,33,49,0.402,22,NaN,—,Dick Motta
3,1969–70,1969–70,—,—,Western,3rd,39,43,0.476,9,Lost Division semifinals (Hawks) 4–1[22],—,Dick Motta
4,1970–71,1970–71,Western,3rd,Midwest,2nd,51,31,0.622,2,Lost conference semifinals (Lakers) 4–3[23],Dick Motta (COY)[6],Dick Motta


### The NLTK Library for Advanced NLP

#### What is NLTK?
**NLTK (Natural Language Toolkit)** is a comprehensive Python library for **natural language processing (NLP)** and **text analysis**. It offers tools for tokenization, preprocessing, part-of-speech tagging, parsing, named-entity recognition, sentiment analysis, and more—useful from classroom demos to research prototypes.


#### Why NLTK Matters in Data Science

| Application Domain        | NLTK Use Cases                                                                 | Business Impact                            |
|--------------------------|----------------------------------------------------------------------------------|--------------------------------------------|
| **Social Media Analytics** | Sentiment, subjectivity, emotion, polarity detection                           | Customer insights, brand monitoring        |
| **Customer Support**       | Text classification, intent detection                                           | Automation, faster resolution              |
| **Content Analysis**       | Keyword extraction, topic modeling, summarization                               | SEO optimization, content strategy         |
| **Research & Academia**    | Corpus analysis, linguistic parsing, concordance tools                          | Rigorous analysis, reproducible studies    |
| **E-commerce**             | Review mining, aspect sentiment, recommendation signals  


#### **💻 NLTK Installation and Setup**

Let's start by installing NLTK and downloading the essential datasets:

In [41]:
pip install nltk

  Using cached joblib-1.5.2-py3-none-any.whl.metadata (5.6 kB)
  Using cached regex-2025.9.18-cp312-cp312-win_amd64.whl.metadata (41 kB)
  Using cached tqdm-4.67.1-py3-none-any.whl.metadata (57 kB)
   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--
   --------------------------- ------------ 1.0/1.5 MB 7.1 MB/s eta 0:00:01
   ---------------------------------------- 1.5/1.5 MB 7.3 MB/s  0:00:00
Using cached regex-2025.9.18-cp312-cp312-win_amd64.whl (275 kB)
Using cached joblib-1.5.2-py3-none-any.whl (308 kB)
Using cached tqdm-4.67.1-py3-none-any.whl (78 kB)

   ---------------------------------------- 0/5 [tqdm]
   -------- ------------------------------- 1/5 [regex]
   ---------------- ----------------------- 2/5 [joblib]
   ---------------- ----------------------- 2/5 [joblib]
   ---------------- ----------------------- 2/5 [joblib]
   ---------------- ----------------------- 2/5 [joblib]
   ------------------------ --------------- 3/5 [click]
   -------------------

In [ ]:
# 🚀 NLTK Installation and Essential Downloads

# Install NLTK (uncomment if not already installed)
# !pip install nltk

try:
    import nltk
    print("✅ NLTK is already installed")
    
    # Download essential datasets (this might take a few minutes first time)
    print("📦 Downloading essential NLTK data...")
    
    # Download essential components
    essential_downloads = [
        'punkt',           # Sentence tokenizer
        'stopwords',       # Stop words for multiple languages  
        'vader_lexicon',   # Sentiment analysis
        'averaged_perceptron_tagger',  # POS tagger
        'wordnet',         # WordNet lexical database
        'omw-1.4',         # Open Multilingual Wordnet
        'movie_reviews',   # Sample movie reviews for sentiment
    ]
    
    for item in essential_downloads:
        try:
            nltk.download(item, quiet=True)
            print(f"  ✓ Downloaded: {item}")
        except:
            print(f"  ⚠️ Could not download: {item}")
    
    print("\n🎯 NLTK Setup Complete!")
    print("Ready for natural language processing!")
    
except ImportError:
    print("❌ NLTK not found. Please install with: pip install nltk")
except Exception as e:
    print(f"⚠️ Setup issue: {e}")
    print("You may need internet connection for first-time downloads")

✅ NLTK is already installed
📦 Downloading essential NLTK data...
  ✓ Downloaded: punkt
  ✓ Downloaded: stopwords
  ✓ Downloaded: vader_lexicon
  ✓ Downloaded: averaged_perceptron_tagger
  ✓ Downloaded: wordnet
  ✓ Downloaded: omw-1.4
  ✓ Downloaded: movie_reviews

🎯 NLTK Setup Complete!
Ready for natural language processing!
  ✓ Downloaded: averaged_perceptron_tagger
  ✓ Downloaded: wordnet
  ✓ Downloaded: omw-1.4
  ✓ Downloaded: movie_reviews

🎯 NLTK Setup Complete!
Ready for natural language processing!


[nltk_data] Error downloading 'vader_lexicon' from
[nltk_data]     <https://raw.githubusercontent.com/nltk/nltk_data/gh-
[nltk_data]     pages/packages/sentiment/vader_lexicon.zip>:
[nltk_data]     <urlopen error [WinError 10054] An existing connection
[nltk_data]     was forcibly closed by the remote host>
[nltk_data] Error downloading 'averaged_perceptron_tagger' from
[nltk_data]     <https://raw.githubusercontent.com/nltk/nltk_data/gh-p
[nltk_data]     ages/packages/taggers/averaged_perceptron_tagger.zip>:
[nltk_data]     <urlopen error [WinError 10054] An existing connection
[nltk_data]     was forcibly closed by the remote host>
[nltk_data] Error downloading 'omw-1.4' from
[nltk_data]     <https://raw.githubusercontent.com/nltk/nltk_data/gh-
[nltk_data]     pages/packages/corpora/omw-1.4.zip>:   <urlopen error
[nltk_data]     [WinError 10054] An existing connection was forcibly
[nltk_data]     closed by the remote host>
[nltk_data] Error downloading 'movie_reviews' from
[nltk_data

#### **Tokenization: Breaking Text into Pieces**

Tokenization is the foundation of NLP - splitting text into meaningful units (words, sentences, or phrases).

In [44]:
# 🔤 Tokenization Examples: Word and Sentence Level
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.tokenize import RegexpTokenizer, WhitespaceTokenizer
import pandas as pd

# Sample text for demonstration
sample_text = """
Natural Language Processing (NLP) is amazing! It helps computers understand human language.
Companies like Google, Amazon, and Microsoft use NLP for various applications.
Did you know that NLTK has been around since 2001? It's a powerful toolkit.
Let's explore tokenization: word-level, sentence-level, and custom patterns.
"""

print("📝 Original Text:")
print(sample_text.strip())

print("\n" + "="*60)
print("🔤 TOKENIZATION EXAMPLES")
print("="*60)

# 1. SENTENCE TOKENIZATION
sentences = sent_tokenize(sample_text)
print(f"\n📄 Sentence Tokenization ({len(sentences)} sentences):")
for i, sentence in enumerate(sentences, 1):
    print(f"  {i}. {sentence.strip()}")

# 2. WORD TOKENIZATION  
words = word_tokenize(sample_text)
print(f"\n🔤 Word Tokenization ({len(words)} tokens):")
print("First 15 tokens:", words[:15])

# 3. CUSTOM TOKENIZATION
# Only alphabetic tokens (removes punctuation and numbers)
alpha_tokenizer = RegexpTokenizer(r'[a-zA-Z]+')
alpha_tokens = alpha_tokenizer.tokenize(sample_text)
print(f"\n🔤 Alphabetic Only ({len(alpha_tokens)} tokens):")
print("First 15 tokens:", alpha_tokens[:15])

# 4. WHITESPACE TOKENIZATION
whitespace_tokenizer = WhitespaceTokenizer()
whitespace_tokens = whitespace_tokenizer.tokenize(sample_text)
print(f"\n🔤 Whitespace Tokenization ({len(whitespace_tokens)} tokens):")
print("First 10 tokens:", whitespace_tokens[:10])

# Create DataFrame for comparison
tokenization_comparison = pd.DataFrame({
    'Method': ['Sentences', 'Words (with punct.)', 'Alphabetic only', 'Whitespace'],
    'Token_Count': [len(sentences), len(words), len(alpha_tokens), len(whitespace_tokens)],
    'Sample_Tokens': [
        [s[:30] + "..." if len(s) > 30 else s for s in sentences[:2]],
        words[:3],
        alpha_tokens[:3], 
        whitespace_tokens[:3]
    ]
})

print(f"\n📊 Tokenization Comparison:")
print(tokenization_comparison.to_string(index=False))

📝 Original Text:
Natural Language Processing (NLP) is amazing! It helps computers understand human language.
Companies like Google, Amazon, and Microsoft use NLP for various applications.
Did you know that NLTK has been around since 2001? It's a powerful toolkit.
Let's explore tokenization: word-level, sentence-level, and custom patterns.

🔤 TOKENIZATION EXAMPLES

📄 Sentence Tokenization (6 sentences):
  1. Natural Language Processing (NLP) is amazing!
  2. It helps computers understand human language.
  3. Companies like Google, Amazon, and Microsoft use NLP for various applications.
  4. Did you know that NLTK has been around since 2001?
  5. It's a powerful toolkit.
  6. Let's explore tokenization: word-level, sentence-level, and custom patterns.

🔤 Word Tokenization (60 tokens):
First 15 tokens: ['Natural', 'Language', 'Processing', '(', 'NLP', ')', 'is', 'amazing', '!', 'It', 'helps', 'computers', 'understand', 'human', 'language']

🔤 Alphabetic Only (48 tokens):
First 15 tokens: 

#### **Text Preprocessing: Stop Words and Cleaning**

Stop words are common words that typically don't carry significant meaning and are often removed to focus on important content.

In [45]:
# 🧹 Stop Words and Text Cleaning
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
import string
import pandas as pd

# Sample product reviews for analysis
product_reviews = [
    "This product is absolutely amazing! I love it so much.",
    "The quality is terrible. I want my money back immediately.",  
    "It's okay, nothing special but does the job well enough.",
    "Fantastic value for money! Highly recommend to everyone.",
    "Poor customer service and the product broke after one day."
]

print("🛍️ Sample Product Reviews:")
for i, review in enumerate(product_reviews, 1):
    print(f"  {i}. {review}")

print("\n" + "="*60)
print("🧹 TEXT PREPROCESSING PIPELINE")
print("="*60)

# Get English stop words
stop_words = set(stopwords.words('english'))
print(f"\n📋 English Stop Words ({len(stop_words)} words):")
print("Sample stop words:", sorted(list(stop_words))[:20])

def preprocess_text(text):
    """Complete text preprocessing pipeline"""
    
    # 1. Convert to lowercase
    text = text.lower()
    
    # 2. Tokenize
    tokens = word_tokenize(text)
    
    # 3. Remove punctuation and non-alphabetic tokens
    tokens = [token for token in tokens if token.isalpha()]
    
    # 4. Remove stop words
    filtered_tokens = [token for token in tokens if token not in stop_words]
    
    return tokens, filtered_tokens

# Process all reviews
processed_reviews = []

print(f"\n🔍 Processing Each Review:")
for i, review in enumerate(product_reviews, 1):
    original_tokens, filtered_tokens = preprocess_text(review)
    
    processed_reviews.append({
        'Review_ID': f'Review_{i}',
        'Original_Text': review,
        'Original_Tokens': len(original_tokens),
        'After_Cleaning': len(filtered_tokens),
        'Tokens_Removed': len(original_tokens) - len(filtered_tokens),
        'Key_Words': filtered_tokens[:5]  # First 5 meaningful words
    })
    
    print(f"\n  📝 Review {i}:")
    print(f"     Original: {review}")
    print(f"     Key words: {filtered_tokens[:8]}")
    print(f"     Tokens: {len(original_tokens)} → {len(filtered_tokens)} (-{len(original_tokens) - len(filtered_tokens)})")

# Create summary DataFrame
df_preprocessing = pd.DataFrame(processed_reviews)

print(f"\n📊 Preprocessing Summary:")
print(df_preprocessing[['Review_ID', 'Original_Tokens', 'After_Cleaning', 'Tokens_Removed']].to_string(index=False))

print(f"\n📈 Overall Statistics:")
total_original = df_preprocessing['Original_Tokens'].sum()
total_cleaned = df_preprocessing['After_Cleaning'].sum()
print(f"• Total tokens before cleaning: {total_original}")
print(f"• Total tokens after cleaning: {total_cleaned}")
print(f"• Tokens removed: {total_original - total_cleaned} ({((total_original - total_cleaned)/total_original)*100:.1f}%)")
print(f"• Average tokens per review (cleaned): {total_cleaned/len(product_reviews):.1f}")

🛍️ Sample Product Reviews:
  1. This product is absolutely amazing! I love it so much.
  2. The quality is terrible. I want my money back immediately.
  3. It's okay, nothing special but does the job well enough.
  4. Fantastic value for money! Highly recommend to everyone.
  5. Poor customer service and the product broke after one day.

🧹 TEXT PREPROCESSING PIPELINE

📋 English Stop Words (198 words):
Sample stop words: ['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any', 'are', 'aren', "aren't", 'as', 'at', 'be', 'because', 'been']

🔍 Processing Each Review:

  📝 Review 1:
     Original: This product is absolutely amazing! I love it so much.
     Key words: ['product', 'absolutely', 'amazing', 'love', 'much']
     Tokens: 10 → 5 (-5)

  📝 Review 2:
     Original: The quality is terrible. I want my money back immediately.
     Key words: ['quality', 'terrible', 'want', 'money', 'back', 'immediately']
     Tokens: 10 → 6 (-4)

  📝 Review 3:
     O

#### **Stemming and Lemmatization: Finding Word Roots**

Both stemming and lemmatization reduce words to their base form, but they work differently:
- **Stemming**: Fast, rule-based chopping (running → run)
- **Lemmatization**: Slower, dictionary-based reduction (better → good)

In [46]:
# 🌱 Stemming vs Lemmatization Comparison
from nltk.stem import PorterStemmer, SnowballStemmer
from nltk.stem import WordNetLemmatizer
import pandas as pd

# Initialize stemmers and lemmatizer
porter = PorterStemmer()
snowball = SnowballStemmer('english')  
lemmatizer = WordNetLemmatizer()

# Test words that demonstrate differences
test_words = [
    'running', 'runs', 'ran', 'runner',        # Run variations
    'better', 'good', 'best',                  # Good variations  
    'dogs', 'cats', 'children', 'geese',      # Plural forms
    'walking', 'walked', 'walks',             # Walk variations
    'flying', 'flies', 'flew',                # Fly variations
    'studies', 'studying', 'studied',         # Study variations
    'beautiful', 'beautifully',               # Adjective/adverb
    'happiness', 'happier', 'happiest'        # Happy variations
]

print("🌱 STEMMING vs LEMMATIZATION COMPARISON")
print("="*60)

# Process each word
results = []
for word in test_words:
    porter_stem = porter.stem(word)
    snowball_stem = snowball.stem(word) 
    lemma = lemmatizer.lemmatize(word)
    lemma_verb = lemmatizer.lemmatize(word, pos='v')  # As verb
    
    results.append({
        'Original': word,
        'Porter_Stem': porter_stem,
        'Snowball_Stem': snowball_stem,
        'Lemma_Noun': lemma,
        'Lemma_Verb': lemma_verb
    })

# Create comparison DataFrame
df_comparison = pd.DataFrame(results)

print("📊 Detailed Comparison:")
print(df_comparison.to_string(index=False))

print(f"\n🎯 Key Differences Analysis:")

# Count unique results for each method
porter_unique = df_comparison['Porter_Stem'].nunique()
snowball_unique = df_comparison['Snowball_Stem'].nunique() 
lemma_unique = df_comparison['Lemma_Noun'].nunique()

print(f"• Porter Stemmer: {porter_unique} unique stems")
print(f"• Snowball Stemmer: {snowball_unique} unique stems") 
print(f"• WordNet Lemmatizer: {lemma_unique} unique lemmas")

print(f"\n💡 Notable Examples:")

# Find interesting cases where methods differ significantly
interesting_cases = [
    ('better', 'good'),   # Lemmatization handles irregular forms
    ('studies', 'study'), # Different handling of -ies ending
    ('flying', 'fly'),    # Verb vs noun treatment
    ('geese', 'goose')    # Irregular plurals
]

for original, expected in interesting_cases:
    if original in test_words:
        row = df_comparison[df_comparison['Original'] == original].iloc[0]
        print(f"• '{original}' → Porter: '{row['Porter_Stem']}', Lemma: '{row['Lemma_Noun']}' (expected: '{expected}')")

# Real-world application example
sample_sentence = "The children were running happily through the beautiful gardens, studying the flying geese."
words_to_process = sample_sentence.lower().replace(',', '').replace('.', '').split()

print(f"\n🔬 Real-world Example:")
print(f"Sentence: {sample_sentence}")
print(f"\nWord processing:")

sentence_results = []
for word in words_to_process:
    if word.isalpha():  # Skip punctuation
        stem = porter.stem(word)
        lemma = lemmatizer.lemmatize(word)
        sentence_results.append(f"{word} → stem: {stem}, lemma: {lemma}")

for result in sentence_results[:6]:  # Show first 6 words
    print(f"  {result}")

print(f"\n⚡ Performance Guidelines:")
print("• Use STEMMING when: Speed is crucial, approximate matching is okay")
print("• Use LEMMATIZATION when: Accuracy is important, working with formal text")
print("• Stemming is ~10x faster but less accurate than lemmatization")

🌱 STEMMING vs LEMMATIZATION COMPARISON
📊 Detailed Comparison:
   Original Porter_Stem Snowball_Stem  Lemma_Noun  Lemma_Verb
    running         run           run     running         run
       runs         run           run         run         run
        ran         ran           ran         ran         run
     runner      runner        runner      runner      runner
     better      better        better      better      better
       good        good          good        good        good
       best        best          best        best        best
       dogs         dog           dog         dog         dog
       cats         cat           cat         cat         cat
   children    children      children       child    children
      geese        gees          gees       goose       geese
    walking        walk          walk     walking        walk
     walked        walk          walk      walked        walk
      walks        walk          walk        walk        walk
     fly

#### **Sentiment Analysis: Understanding Emotions**

NLTK provides powerful tools for analyzing sentiment - determining whether text expresses positive, negative, or neutral emotions.

In [48]:
# 📊 Sentiment Analysis with NLTK VADER
from nltk.sentiment import SentimentIntensityAnalyzer
import pandas as pd
import matplotlib.pyplot as plt
nltk.download('vader_lexicon')

# Initialize VADER sentiment analyzer
analyzer = SentimentIntensityAnalyzer()

# Diverse sample texts for sentiment analysis
sample_texts = [
    # Positive examples
    "This product is absolutely amazing! I love everything about it!",
    "Fantastic customer service and fast delivery. Highly recommend!",
    "Great value for money. Very satisfied with my purchase.",
    
    # Negative examples  
    "Terrible quality, waste of money. Very disappointed.",
    "Poor customer service and the product broke immediately.",
    "Worst purchase I've ever made. Completely useless.",
    
    # Neutral examples
    "The product arrived on time. It's okay, nothing special.",
    "Standard quality for the price. Does what it's supposed to do.",
    "Average product with typical features. Works as expected.",
    
    # Mixed sentiment
    "Great design but poor durability. Mixed feelings about this.",
    "Love the features but hate the price. It's complicated.",
]

print("📊 SENTIMENT ANALYSIS WITH NLTK VADER")
print("="*60)

# Analyze sentiment for each text
sentiment_results = []

for i, text in enumerate(sample_texts, 1):
    # Get sentiment scores
    scores = analyzer.polarity_scores(text)
    
    # Determine overall sentiment
    if scores['compound'] >= 0.05:
        overall = 'Positive'
        emoji = '😊'
    elif scores['compound'] <= -0.05:
        overall = 'Negative'  
        emoji = '😞'
    else:
        overall = 'Neutral'
        emoji = '😐'
    
    sentiment_results.append({
        'Text_ID': f'Text_{i:02d}',
        'Text': text[:50] + "..." if len(text) > 50 else text,
        'Positive': round(scores['pos'], 3),
        'Neutral': round(scores['neu'], 3), 
        'Negative': round(scores['neg'], 3),
        'Compound': round(scores['compound'], 3),
        'Sentiment': overall,
        'Emoji': emoji,
        'Full_Text': text
    })
    
    print(f"{emoji} Text {i}: {overall} (compound: {scores['compound']:.3f})")
    print(f"   \"{text}\"")
    print(f"   Pos: {scores['pos']:.2f}, Neu: {scores['neu']:.2f}, Neg: {scores['neg']:.2f}")
    print()

# Create DataFrame for analysis
df_sentiment = pd.DataFrame(sentiment_results)

print("📈 SENTIMENT ANALYSIS SUMMARY")
print("="*40)

# Overall statistics
sentiment_counts = df_sentiment['Sentiment'].value_counts()
print("📊 Sentiment Distribution:")
for sentiment, count in sentiment_counts.items():
    percentage = (count / len(df_sentiment)) * 100
    print(f"  {sentiment}: {count} texts ({percentage:.1f}%)")

print(f"\n📏 Score Ranges:")
print(f"• Positive scores: {df_sentiment['Positive'].min():.3f} - {df_sentiment['Positive'].max():.3f}")
print(f"• Negative scores: {df_sentiment['Negative'].min():.3f} - {df_sentiment['Negative'].max():.3f}")
print(f"• Compound scores: {df_sentiment['Compound'].min():.3f} - {df_sentiment['Compound'].max():.3f}")

# Show most extreme examples
print(f"\n🎯 Most Extreme Examples:")
most_positive = df_sentiment.loc[df_sentiment['Compound'].idxmax()]
most_negative = df_sentiment.loc[df_sentiment['Compound'].idxmin()]

print(f"😊 Most Positive (score: {most_positive['Compound']:.3f}):")
print(f"   \"{most_positive['Full_Text']}\"")
print(f"😞 Most Negative (score: {most_negative['Compound']:.3f}):")
print(f"   \"{most_negative['Full_Text']}\"")

# Business application example
print(f"\n💼 Business Application Example:")
print("This could be used for:")
print("• Customer review analysis")
print("• Social media monitoring") 
print("• Product feedback categorization")
print("• Brand sentiment tracking")
print("• Customer support prioritization")

# Display summary table
print(f"\n📋 Detailed Results:")
display_df = df_sentiment[['Text_ID', 'Text', 'Compound', 'Sentiment', 'Emoji']].copy()
print(display_df.to_string(index=False))

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\lsi8012\AppData\Roaming\nltk_data...


📊 SENTIMENT ANALYSIS WITH NLTK VADER
😊 Text 1: Positive (compound: 0.879)
   "This product is absolutely amazing! I love everything about it!"
   Pos: 0.57, Neu: 0.43, Neg: 0.00

😊 Text 2: Positive (compound: 0.771)
   "Fantastic customer service and fast delivery. Highly recommend!"
   Pos: 0.53, Neu: 0.47, Neg: 0.00

😊 Text 3: Positive (compound: 0.862)
   "Great value for money. Very satisfied with my purchase."
   Pos: 0.61, Neu: 0.39, Neg: 0.00

😞 Text 4: Negative (compound: -0.852)
   "Terrible quality, waste of money. Very disappointed."
   Pos: 0.00, Neu: 0.30, Neg: 0.70

😞 Text 5: Negative (compound: -0.710)
   "Poor customer service and the product broke immediately."
   Pos: 0.00, Neu: 0.50, Neg: 0.50

😞 Text 6: Negative (compound: -0.802)
   "Worst purchase I've ever made. Completely useless."
   Pos: 0.00, Neu: 0.41, Neg: 0.59

😞 Text 7: Negative (compound: -0.092)
   "The product arrived on time. It's okay, nothing special."
   Pos: 0.17, Neu: 0.63, Neg: 0.20

😐 Text 8: N

#### **Part-of-Speech (POS) Tagging & Named Entity Recognition**

POS tagging identifies grammatical roles of words (noun, verb, adjective), while Named Entity Recognition (NER) finds important entities like names, places, and organizations.

In [54]:
# 🏷️ POS Tagging and Named Entity Recognition
import nltk
from nltk.tokenize import word_tokenize
from nltk.tag import pos_tag
from nltk.chunk import ne_chunk
from nltk.tree import Tree
nltk.download('averaged_perceptron_tagger_eng')
nltk.download('maxent_ne_chunker_tab')
nltk.download('maxent_ne_chunker_tab')
nltk.download('words')
nltk.download('words')


import pandas as pd

# Sample business-related text
business_text = """
Apple Inc. is planning to open a new headquarters in Austin, Texas next year.
CEO Tim Cook announced this during a meeting with investors on Wall Street.
The company expects to hire 5,000 new employees, including software engineers 
and data scientists. Microsoft and Google are also expanding their operations
in the region. The project will cost approximately $1 billion dollars.
"""

print("🏷️ POS TAGGING & NAMED ENTITY RECOGNITION")
print("="*60)

print("📝 Sample Text:")
print(business_text.strip())

# 1. POS TAGGING
print(f"\n🔤 PART-OF-SPEECH TAGGING:")
tokens = word_tokenize(business_text)
pos_tags = pos_tag(tokens)

# Group by POS types for analysis
pos_groups = {}
for word, pos in pos_tags:
    if pos not in pos_groups:
        pos_groups[pos] = []
    pos_groups[pos].append(word)

print(f"Found {len(set([pos for _, pos in pos_tags]))} different POS tags:")

# Show most common POS categories
pos_description = {
    'NN': 'Noun (singular)',
    'NNS': 'Noun (plural)', 
    'NNP': 'Proper noun (singular)',
    'NNPS': 'Proper noun (plural)',
    'VB': 'Verb (base form)',
    'VBD': 'Verb (past tense)',
    'VBG': 'Verb (gerund/present participle)',
    'VBN': 'Verb (past participle)',
    'VBP': 'Verb (present, not 3rd person singular)',
    'VBZ': 'Verb (present, 3rd person singular)',
    'JJ': 'Adjective',
    'JJR': 'Adjective (comparative)', 
    'JJS': 'Adjective (superlative)',
    'RB': 'Adverb',
    'DT': 'Determiner',
    'IN': 'Preposition/subordinating conjunction',
    'TO': 'to',
    'CD': 'Cardinal number'
}

for pos in sorted(pos_groups.keys()):
    if pos in pos_description:
        example_words = pos_groups[pos][:3]  # Show first 3 examples
        print(f"  {pos} ({pos_description[pos]}): {example_words}")

# 2. NAMED ENTITY RECOGNITION  
print(f"\n🏢 NAMED ENTITY RECOGNITION:")
named_entities = ne_chunk(pos_tags)

# Extract entities
entities = []
def extract_entities(tree):
    for subtree in tree:
        if isinstance(subtree, Tree):  # It's a named entity
            entity_name = ' '.join([token for token, pos in subtree.leaves()])
            entity_label = subtree.label()
            entities.append((entity_name, entity_label))
            
extract_entities(named_entities)

if entities:
    print("🎯 Detected Entities:")
    entity_types = {}
    for entity, label in entities:
        if label not in entity_types:
            entity_types[label] = []
        entity_types[label].append(entity)
    
    for label, entity_list in entity_types.items():
        print(f"  {label}: {entity_list}")
else:
    print("  No named entities detected with basic NER")

# 3. MANUAL ENTITY EXTRACTION (for demonstration)
print(f"\n🔍 MANUAL ENTITY IDENTIFICATION:")

# Find potential entities using POS tags
companies = [word for word, pos in pos_tags if pos == 'NNP' and word in ['Apple', 'Microsoft', 'Google']]
people = [word for word, pos in pos_tags if pos == 'NNP' and word in ['Tim', 'Cook']]
locations = [word for word, pos in pos_tags if pos == 'NNP' and word in ['Austin', 'Texas', 'Wall', 'Street']]
numbers = [word for word, pos in pos_tags if pos == 'CD']

manual_entities = {
    'Companies': companies,
    'People': people,
    'Locations': locations, 
    'Numbers': numbers
}

for category, items in manual_entities.items():
    if items:
        print(f"  {category}: {items}")

# 4. BUSINESS APPLICATIONS
print(f"\n💼 BUSINESS APPLICATIONS:")
print("POS Tagging can be used for:")
print("• Information extraction from documents")
print("• Improving search relevance")  
print("• Text summarization")
print("• Grammar checking tools")
print("• Keyword extraction")

print("\nNamed Entity Recognition can be used for:")
print("• Customer relationship management")
print("• Compliance and regulatory analysis")
print("• News and social media monitoring")
print("• Document classification")
print("• Knowledge graph construction")

# Create summary DataFrame
word_analysis = []
for word, pos in pos_tags[:15]:  # First 15 words
    if word.isalpha():  # Skip punctuation
        is_entity = any(word in entity for entity, _ in entities)
        word_analysis.append({
            'Word': word,
            'POS_Tag': pos,
            'POS_Description': pos_description.get(pos, 'Other'),
            'Is_Entity': is_entity
        })

df_analysis = pd.DataFrame(word_analysis)
print(f"\n📋 Sample Word Analysis:")
print(df_analysis.to_string(index=False))

[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     C:\Users\lsi8012\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package maxent_ne_chunker_tab to
[nltk_data]     C:\Users\lsi8012\AppData\Roaming\nltk_data...
[nltk_data]   Package maxent_ne_chunker_tab is already up-to-date!
[nltk_data] Downloading package maxent_ne_chunker_tab to
[nltk_data]     C:\Users\lsi8012\AppData\Roaming\nltk_data...
[nltk_data]   Package maxent_ne_chunker_tab is already up-to-date!
[nltk_data] Downloading package words to
[nltk_data]     C:\Users\lsi8012\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\words.zip.
[nltk_data] Downloading package words to
[nltk_data]     C:\Users\lsi8012\AppData\Roaming\nltk_data...
[nltk_data]   Package words is already up-to-date!
[nltk_data] Downloading package words to
[nltk_data]     C:\Users\lsi8012\AppData\Roaming\nltk_data...
[n

🏷️ POS TAGGING & NAMED ENTITY RECOGNITION
📝 Sample Text:
Apple Inc. is planning to open a new headquarters in Austin, Texas next year.
CEO Tim Cook announced this during a meeting with investors on Wall Street.
The company expects to hire 5,000 new employees, including software engineers 
and data scientists. Microsoft and Google are also expanding their operations
in the region. The project will cost approximately $1 billion dollars.

🔤 PART-OF-SPEECH TAGGING:
Found 20 different POS tags:
  CD (Cardinal number): ['5,000', '1', 'billion']
  DT (Determiner): ['a', 'this', 'a']
  IN (Preposition/subordinating conjunction): ['in', 'during', 'with']
  JJ (Adjective): ['new', 'next', 'new']
  NN (Noun (singular)): ['headquarters', 'year', 'meeting']
  NNP (Proper noun (singular)): ['Apple', 'Inc.', 'Austin']
  NNS (Noun (plural)): ['investors', 'employees', 'engineers']
  RB (Adverb): ['also', 'approximately']
  TO (to): ['to', 'to']
  VB (Verb (base form)): ['open', 'hire', 'cost']
  VBD (

#### **NLTK Summary & Real-World Applications**

##### **📊 NLTK Capabilities Overview**

| Feature | Purpose | Business Use Case | Performance |
|---------|---------|------------------|-------------|
| **Tokenization** | Split text into words/sentences | Document processing, search | ⚡ Fast |
| **Stop Words** | Remove common words | Focus on meaningful content | ⚡ Fast |  
| **Stemming** | Reduce words to stems | Search optimization, indexing | ⚡ Very Fast |
| **Lemmatization** | Reduce words to dictionary form | Text analysis, normalization | 🔄 Medium |
| **Sentiment Analysis** | Detect emotions and opinions | Customer feedback, social monitoring | ⚡ Fast |
| **POS Tagging** | Identify grammatical roles | Information extraction | 🔄 Medium |
| **Named Entity Recognition** | Find people, places, organizations | Document analysis, CRM | 🔄 Medium |


##### **💡 Best Practices & Tips**

**Performance Optimization:**

- Use **stemming** for speed-critical applications  
- Use **lemmatization** for accuracy-critical applications
- **Cache** processed results for repeated text analysis
- **Batch process** large datasets instead of one-by-one

**Data Quality:**

- **Clean text** before processing (remove HTML, special characters)
- **Handle multiple languages** appropriately
- **Validate results** with domain experts
- **Consider context** when interpreting sentiment


NLTK provides an excellent foundation for understanding NLP concepts before moving to more advanced tools!